### PR TITLE
Glacier migration

### PR DESCRIPTION
**Pull request by srchase**
migrates and applies AddGlacierApiVersionMiddleware

For now, this hard-codes the api version. This will need to be updated with a way to dynamically pull in the version for the service.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 1]
```diff
@@ -67,6 +67,11 @@
                         .withConventions("@aws-sdk/middleware-expect-continue", "^0.1.0-preview.5",
                                         "AddExpectContinue", HAS_MIDDLEWARE)
                         .servicePredicate((m, s) -> s.getId().getName().equals("AmazonS3"))
+                        .build(),
+                RuntimeClientPlugin.builder()
+                                .withConventions("@aws-sdk/middleware-sdk-glacier", "^0.1.0-preview.7",
+                                                "AddGlacierApiVersion", HAS_MIDDLEWARE)
+                                .servicePredicate((m, s) -> s.getId().getName().equals("Glacier"))
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: Probably safer to use the service ID here.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 2]
```diff
@@ -67,6 +67,11 @@
                         .withConventions("@aws-sdk/middleware-expect-continue", "^0.1.0-preview.5",
                                         "AddExpectContinue", HAS_MIDDLEWARE)
                         .servicePredicate((m, s) -> s.getId().getName().equals("AmazonS3"))
+                        .build(),
+                RuntimeClientPlugin.builder()
+                                .withConventions("@aws-sdk/middleware-sdk-glacier", "^0.1.0-preview.7",
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: Nit: Can you line up the indentation with the others?

Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in cd4c2b99686d722ba015aae49e75a7b5e72e7067

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java (hunk 1 above)
Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in 65c11d13d78988b8c00fa594899ff29bd0046122

File: packages/middleware-sdk-glacier/src/add-glacier-api-version.ts [hunk 3]
```diff
@@ -1,18 +1,43 @@
...
-import { headerDefault } from "@aws-sdk/middleware-header-default";
+export function addGlacierApiVersionMiddleware(): BuildMiddleware<any, any> {
+  return <Output extends MetadataBearer>(
+    next: BuildHandler<any, Output>
+  ): BuildHandler<any, Output> => async (
+    args: BuildHandlerArguments<any>
+  ): Promise<BuildHandlerOutput<Output>> => {
+    let request = args.request;
+    if (HttpRequest.isInstance(request)) {
+      request.headers["x-amz-glacier-version"] = "2012-06-01";
```
Comment (AllanFly120) [inline packages/middleware-sdk-glacier/src/add-glacier-api-version.ts]: This probably need to change after #490 lands

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 4]
```diff
@@ -56,18 +58,27 @@
...
+                        .servicePredicate((m, s) -> testServiceId(s, "S3"))
                         .build(),
                 RuntimeClientPlugin.builder()
                         .withConventions("@aws-sdk/middleware-expect-continue", "^0.1.0-preview.5",
                                         "AddExpectContinue", HAS_MIDDLEWARE)
-                        .servicePredicate((m, s) -> s.getId().getName().equals("AmazonS3"))
+                        .servicePredicate((m, s) -> testServiceId(s, "S3"))
+                        .build(),
+                RuntimeClientPlugin.builder()
+                        .withConventions("@aws-sdk/middleware-sdk-glacier", "^0.1.0-preview.7",
```
Comment (AllanFly120) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: One more thing: Can you add these middleware to `AwsDependency` class, this way we can easier maintain the dependencies

Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in 8766e3ef7341a929b8c5c395b2e2e55a509e6682

**Pull request merged**